api/provider/aws/service_list.go

```diff
@@ -1,110 +1,81 @@
 	}
 
-	clusterServices := map[*string][]*string{}
-
-	for _, clusterArn := range clusterArns {
-		clusterName := strings.Split(aws.StringValue(clusterArn), ":")[5]
-		clusterName = strings.Replace(clusterName, "cluster/", "", -1)
-
-		if !strings.HasPrefix(clusterName, s.Config.Instance()) {
-			continue
-		}
-
-		serviceArns, err := s.listServices(clusterArn)
+	serviceIDs := make([]string, len(serviceARNs))
+	for i, serviceARN := range serviceARNs {
+		serviceID, err := lookupServiceIDFromServiceARN(s.TagStore, serviceARN)
 		if err != nil {
 			return nil, err
 		}
 
-		clusterServices[clusterArn] = serviceArns
-	}
-
-	serviceSummaries := []models.ServiceSummary{}
-	for _, serviceArns := range clusterServices {
-		for _, serviceArn := range serviceArns {
-			ecsServiceID := s.serviceARNToECSServiceID(aws.StringValue(serviceArn))
-			service := &models.Service{
-				ServiceID: ecsServiceID,
-			}
-			s.populateSummariesTags(service, ecsServiceID)
-
-			summary := models.ServiceSummary{
-				ServiceID:       service.ServiceID,
-				ServiceName:     service.ServiceName,
-				EnvironmentID:   service.EnvironmentID,
-				EnvironmentName: service.EnvironmentName,
-			}
-
-			serviceSummaries = append(serviceSummaries, summary)
-		}
+		serviceIDs[i] = serviceID
 	}
 
-	return serviceSummaries, nil
+	return s.makeServiceSummaryModels(serviceIDs)
 }
 
-func (s *ServiceProvider) listServices(clusterArn *string) ([]*string, error) {
-	serviceArns := []*string{}
-	input := &ecs.ListServicesInput{
-		Cluster: clusterArn,
-	}
-
-	for {
-		output, err := s.AWS.ECS.ListServices(input)
-		if err != nil {
-			return nil, err
+func (s *ServiceProvider) listServiceARNs() ([]string, error) {
+	var serviceARNs []string
+	fn := func(output *ecs.ListServicesOutput, lastPage bool) bool {
+		for _, serviceARN := range output.ServiceArns {
+			serviceARNs = append(serviceARNs, aws.StringValue(serviceARN))
 		}
 
-		serviceArns = append(serviceArns, output.ServiceArns...)
-		input.NextToken = output.NextToken
+		return !lastPage
+	}
 
-		if output.NextToken == nil {
-			break
-		}
+	if err := s.AWS.ECS.ListServicesPages(&ecs.ListServicesInput{}, fn); err != nil {
+		return nil, err
 	}
 
-	return serviceArns, nil
+	return serviceARNs, nil
 }
 
-func (s *ServiceProvider) listClusters() ([]*string, error) {
-	input := &ecs.ListClustersInput{}
-	clusterArns := []*string{}
+func (s *ServiceProvider) makeServiceSummaryModels(serviceIDs []string) ([]models.ServiceSummary, error) {
+	serviceTags, err := s.TagStore.SelectByType("service")
+	if err != nil {
+		return nil, err
+	}
+
+	models := make([]models.ServiceSummary, len(serviceIDs))
+	for i, serviceID := range serviceIDs {
+		models[i].ServiceID = serviceID
+
+		if tag, ok := serviceTags.WithID(serviceID).WithKey("name").First(); ok {
+			models[i].ServiceName = tag.Value
+		}
 
-	for {
-		output, err := s.AWS.ECS.ListClusters(input)
+		environmentID, err := lookupEntityEnvironmentID(s.TagStore, "service", serviceID)
```

Comment:
I don't think we need this helper function. In each case it would be used (as far as I've seen), we already have the tags that contain the `entity_id` tag. It's way faster to iterate over the tags we already have than to burden the db with another select operation.

Reply:
✅